README.txt
Ni Karma System
Designed by Vuelhering and Qed of Icecrown (stef+nks@swcp-dot-com)
Plugin originally written by Mavios of Icecrown, with extensive changes by Vuelhering.

Read and understand the system.  This plugin helps you implement the system, but it cannot do it automatically.  You *must* understand how it works or this plugin will do you no good, so look in the Documentation folder.  The way it works looks scary at first, but it's fairly easy to understand once you figure it out.

Install the Ni_Karma and Ni_Loader folders in <wowdir>/Interface/AddOns and make sure that you see it on the addons button at the character select screen.

If your website is capable of running php, install the karma.php and karma.css scripts on your website.  After running the plugin and assigning loot, the loot master must upload the Karma data to the same directory as the php script on the website.  The datafile to upload is <wowdir>/WTF/Account/<login>/SavedVariables/Ni_Karma.lua
This is where WoW automatically saves variables, which are written out when the game exits, and loaded when the game starts.

After it's uploaded, you can access the NKS info by going to http://www.yourwebsite.com/karma.php

Note - DO NOT ASK ME about your website... ask your ISP if you have questions about running php scripts on your webserver.




-- Copyright 2006-2009, Mavios and Vuelhering, Knights who say Ni, Icecrown
-- 
-- Permission granted for use, modification, and distribution provided:
-- 1. Any distributions include the original distribution in its entirety, OR a working URL to freely get the entire original distribution is clearly listed in the documentation.
-- 2. Any modified distributions clearly mark YOUR changes, or document the changes somehow.
-- 3. Any modified distributions MUST NOT imply in any way that it is an official upgrade version of this software (such as NKS+ or Enhanced Karma System, or probably anything with "NKS" or "Karma" in the name).  If you want your changes in the official distribution, write (stef+nks @swcp.com) and it might get included.
-- 4. No fee is charged for any distribution of this software (modified or original).
-- 
-- Snippets of code "borrowed" (fewer than 100 total lines) can merely include the URL http://www.knights-who-say-ni.com/NKS and credit for the code used.
-- The Ni_Karma.toc file is granted to the public domain, and may be changed as desired.

Changes by jameskilton:

 - Alt-Clicking starts the bidding instead of just clicking
 - Hooks into the WoW Loot mod for class priority


Changes.txt
3.0 r03 added Loading on demand.  It will load when entering a lockout dungeon in a raid.  Requires the Ni_Loader addon, included.

3.0 r02 added deathknights, and some german localizations.

3.0 r01 updated some of the functions to handle WotFLK

2.4 r01 added German functional translation (will handle rolls).  Freezes item when populating roll window, added button to clear current item.

2.2 r01 fixed some issues with the 2.2 changes.  Website scripts now point to wow armory for item info instead of allakhazam.

2.1 r01 fixed some issues with the 2.1 changes.

2.0 r03 New karma.php web script to handle changes in savedvariables.

2.0 r01 Renumbered based on WoW version, updated to run with WoW 2.0.  Some minor enhancements.

0.93.1 Added data format tag to help upgrading.  Added lifetime karma, which is all points added up, but ignoring points lost from items.  More bugfixes.  Added /km create command.  Changed web scripts to read lifetime karma.

0.92.3 beta - more bugfixes

0.92.2 beta - important bugfix (open rolls were breaking db)

0.92.1 beta - Major rewrite to allow localizations.  Fixed parser functions, fixed compact routine, other stuff.  Need translations.  Probably still bugs, so save your old backup karma files... just in case :-)

0.92 beta - This has some bugfixes and a minor change where "nobonus" declarations on a class item will use up to 50 points of karma. This is because there is a minimum cost of 25 (generally), and had they used bonus with 50 points, it would cost them the same 25. This makes it a little friendlier to long-time raiders. Also added a compression function in case your website php scripts are running out of memory. Note that if you compress the file, it will NOT BE BACKWARD COMPATIBLE. You must not use a previous version on the new file (or it'll overwrite entries, although the final karma will be correct).  Still needs work on compression algorithm.

0.91.1 - updated TOC (11000)

0.91 - bugfixes to initial 0.90


NKS_Roll.lua
-- Ni Karma System (NKS) for raid loot distribution
-- The Ni Karma System was designed by Vuelhering (stef+nks @swcp.com) and Qed of Icecrown
-- 
-- Plugin coded by Mavios of Icecrown (althar @gmail.com).  Thanks Mavios, you rock.
-- Code maintenance and additional programming by Mavios and Vuelhering
-- Instructions for use at http://www.knights-who-say-ni.com/NKS
-- 
-- Copyright 2006-2009, Mavios and Vuelhering, Knights who say Ni, Icecrown
-- 
-- Permission granted for use, modification, and distribution provided:
-- 1. Any distributions include the original distribution in its entirety, OR a working URL to freely get the entire original distribution is clearly listed in the documentation.
-- 2. Any modified distributions clearly mark YOUR changes, or document the changes somehow.
-- 3. Any modified distributions MUST NOT imply in any way that it is an official upgrade version of this software (such as NKS+ or Enhanced Karma System, or probably anything with "NKS" or "Karma" in the name).  If you want your changes in the official distribution, write (stef+nks @swcp.com) and it might get included.
-- 4. No fee is charged for any distribution of this software (modified or original).
-- 
-- Snippets of code "borrowed" (fewer than 100 total lines) can merely include the URL http://www.knights-who-say-ni.com/NKS and credit for the code used.


---------------------------------------
----    Karma Roll Window Routines ----
---------------------------------------
function KarmaRoll_OnLoad()
    this:RegisterEvent("ADDON_LOADED"); 
    this:RegisterEvent("CHAT_MSG_SYSTEM"); 
    this:RegisterEvent("CHAT_MSG_WHISPER"); 

    hooksecurefunc("HandleModifiedItemClick", KarmaLootItem_OnClick)
    hooksecurefunc("SetItemRef", Karma_SetItemRef) 
end

-- Main event handler
function KarmaRoll_OnEvent(event)

    if (event == "ADDON_LOADED") then
	KarmaRollFrameItem:SetText("");
	KarmaRollFrameBaseCost:SetText("0");
	KarmaRollFrameBaseCost:SetNumeric(1);
	KarmaRollFrameFinalKarma:SetText("0");
	KarmaRollFrameFinalKarma:SetNumeric(1);

    elseif (nks.Active and event == "CHAT_MSG_WHISPER") then
	local cmd, extra = Karma_GetToken(arg1);
	cmd = Karma_StripTok(cmd);
	local cmd2, extra2 = Karma_GetToken(extra);
	cmd2 = Karma_StripTok(cmd2);
	if (cmd == nks.KMSG.BONUS or cmd == nks.KMSG.NOBONUS or (cmd == nks.KMSG.NOBONUS1 and cmd2 == nks.KMSG.NOBONUS2)) then
	    KarmaRoll_AddPlayer(arg2, cmd == nks.KMSG.BONUS);
	end
    elseif (nks.Active and event == "CHAT_MSG_SYSTEM" and string.find(arg1, nks.KMSG.SYS.ROLLS) and string.find(arg1, "%(1%-100%)")) then 
	_, _, player_name, player_roll = string.find(arg1, "(.+) " .. nks.KMSG.SYS.ROLLS .. " (%d+)");
	if (player_name ~= nil and player_roll ~= nil) then
	    KarmaRoll_Roll(player_name, player_roll);
	end
    end 
end


-- A player declared intention to roll, add them to list
function KarmaRoll_AddPlayer(player_name, use_bonus)
    local player, fullname, class = Karma_Getstats(player_name);

    if (KarmaList[Raid_Name][player] == nil) then
	-- Add player to system
	Karma_Newplayer(player);
    end

    -- Check if they are already in list  
    for i=1, #(nks.RollList) do
	if (nks.RollList[i][1] == fullname) then
	    if (nks.RollList[i][4] > 0) then
		Karma_message(nks.KMSG.PLAYER .. fullname .. nks.KMSG.ROLLED .. tostring(nks.RollList[i][4]) .. nks.KMSG.REROLLED);
	    end
	    table.remove(nks.RollList, i);
	    break;
	end
    end

    if (class) then
	-- Player is in the raid if we found their class
	if (KarmaList[Raid_Name][player]["points"] < 0) then
	    use_bonus = true; -- always bonus if you're negative muahhaa
	end

	table.insert(nks.RollList, {fullname, class, 0, 0, 0, use_bonus});
	if (not nks.OpenRoll) then
	    if (use_bonus) then
		Karma_message(nks.KMSG.REPLYBONUS1 .. KarmaList[Raid_Name][player]["points"] .. nks.KMSG.REPLYBONUS2, nks.KARMA_SHOWTO_PLAYER, player);
	    else
		Karma_message(nks.KMSG.REPLYNOBONUS, nks.KARMA_SHOWTO_PLAYER, player);
	    end
	end
	KarmaRollList_Update();

    else
	Karma_message(nks.KMSG.PLAYER .. player_name .. nks.KMSG.NOTINRAID);
    end
end

-- A /roll was detected, update the list if they are in it

function KarmaRoll_Roll(player_name, player_roll)
    local player=string.lower(player_name);
-- If it's an open (no bonus declaration required) roll then add them to the list when they /roll
    if (nks.OpenRoll) then
	if (KarmaList[Raid_Name][player] == nil) then
	    Karma_Newplayer(player_name);
	end
	KarmaRoll_AddPlayer(player_name, false)
    end

    for i=1, #(nks.RollList) do
	if (string.lower(nks.RollList[i][1]) == player) then
	    if (nks.RollList[i][4] > 0) then
		Karma_message(nks.KMSG.PLAYER .. player_name .. nks.KMSG.ROLLED .. tostring(nks.RollList[i][4]) .. nks.KMSG.REROLLED);
	    else
		nks.RollList[i][4] = tonumber(player_roll);
	    end
	    nks.RollList[i][5] = nks.RollList[i][3] + nks.RollList[i][4];
	    break;
	end
    end

    KarmaRollList_Update();
end

-- Sort the rolls
function KarmaRoll_Sort(a, b)
    if (a[5] == b[5]) then
      return (a[1] < b[1]);
    else
      return (a[5] > b[5]);
    end;
end

-- Update the display
function KarmaRollList_Update()

    local rollOffset = FauxScrollFrame_GetOffset(KarmaRollScrollFrame);
    local base_cost = tonumber(KarmaRollFrameBaseCost:GetText());
    if (base_cost == nil) then
	base_cost = 0;
    end

    local numRolls = #(nks.RollList);
    local min_deduction = KarmaConfig["MIN_KARMA_NONCLASS_DEDUCTION"];
    local max_deduction = KarmaConfig["MAX_KARMA_NONCLASS_DEDUCTION"];

    --  (RollList, {player_name, class, karma_used, 0, karma_used, use_bonus});

    -- "nobonus" will use 2 * base cost of the item.  It will also force using bonus if
    -- your karma is less than this value, since it'll cost you anyway.
    -- ex: you have 20 pts on a 25-point class item.  It will force you to use bonus of 20 pts no matter what.
    -- ex: you have 120 pts, nobonusing on a 25-point class item.  It will use 50 karma (or 2x the min cost).
    -- ex: you have 120 pts, nobonusing a non-class item.  It will use 0 karma.

    for i=1, numRolls do
	local player,karma_used,use_bonus = nks.RollList[i][1], nks.RollList[i][3], nks.RollList[i][6];
	player = string.lower(player);
	if (nks.OpenRoll) then
	    karma_used = 0;
	elseif (use_bonus) then
	    karma_used = KarmaList[Raid_Name][player]["points"];
	else
	    karma_used = min(2 * base_cost , KarmaList[Raid_Name][player]["points"]);
	end

	nks.RollList[i][3] = karma_used;
	nks.RollList[i][5] = nks.RollList[i][3] + nks.RollList[i][4];
    end
    
    table.sort(nks.RollList, KarmaRoll_Sort);
    
    for i=1, nks.ROLLS_TO_DISPLAY do
	local rollIndex = rollOffset + i;
	button = getglobal("KarmaRollFrameButton"..i);
	button.rollIndex = rollIndex;
	local roll_info = nks.RollList[rollIndex];
	if (roll_info ~= nil) then
	    getglobal("KarmaRollFrameButton"..i.."Name"):SetText(roll_info[1]);
	    getglobal("KarmaRollFrameButton"..i.."Class"):SetText(roll_info[2]);
	    getglobal("KarmaRollFrameButton"..i.."Karma"):SetText(roll_info[3]);
	    getglobal("KarmaRollFrameButton"..i.."Roll"):SetText(roll_info[4]);
	    getglobal("KarmaRollFrameButton"..i.."Total"):SetText(roll_info[5]);
	end

	-- Lock the highlight if a roller is selected
	if ( KarmaRollFrame.selectedRoller == rollIndex ) then
	    button:LockHighlight();

	-- Here are the guts: divide the karma by 2
	-- (and drop fractions to the nearest rounding value)
	-- This is the BASIS of the zero-sum property
	    local final_value = ceil(roll_info[3] / 2 / KarmaConfig["KARMA_ROUNDING"]) * KarmaConfig["KARMA_ROUNDING"];
	    local base_cost = tonumber(KarmaRollFrameBaseCost:GetText());
	    if (base_cost == nil) then
		base_cost = 0;
	    end
	    if (final_value < base_cost) then
		KarmaRollFrameFinalKarma:SetText(base_cost);
	    else
		if (max_deduction > 0 and max_deduction < final_value) then
		    final_value = max_deduction;
		end
		KarmaRollFrameFinalKarma:SetText(final_value);
	    end
	else
	    button:UnlockHighlight();
	end

	if ( rollIndex > numRolls ) then
	    button:Hide();
	else
	    button:Show();
	end
    end  
    
    -- Enable/disable buttons
    if (KarmaRollFrame.selectedRoller ~= nil and KarmaRollFrame.selectedRoller > 0 and
			KarmaRollFrameItem:GetText() ~= "") then
	KarmaRollFrameAwardButton:Enable();
    else
	KarmaRollFrameAwardButton:Disable();
    end
    
    -- ScrollFrame stuff
    FauxScrollFrame_Update(KarmaRollScrollFrame, numRolls, nks.ROLLS_TO_DISPLAY, nks.ROLL_FRAME_ROLL_HEIGHT );

end

-- Generic click handler for roller list
function KarmaRollFrameRollButton_OnClick(button)

    if ( button == "LeftButton" ) then
	KarmaRollFrame.selectedRoller = getglobal("KarmaRollFrameButton"..this:GetID()).rollIndex;
	KarmaRollList_Update();
    else
	-- Remove from list
	local roller = getglobal("KarmaRollFrameButton"..this:GetID()).rollIndex
	table.remove(nks.RollList, roller);

    -- If we removed the last one, move bar up
    if (KarmaRollFrame.selectedRoller > #(nks.RollList)) then
	KarmaRollFrame.selectedRoller = #(nks.RollList);
    end
    KarmaRollList_Update();
  end
  
end

-- Clear the rollers list
function KarmaRollFrame_Clear()
    nks.RollList = {};
    KarmaRollFrame.selectedRoller = 0;
    Karma_Clearitem();
    nks.OpenRoll = false;
    KarmaRollFrameOpenCheckButton:SetChecked(nks.OpenRoll);
    KarmaRollList_Update();
end

-- Declare winner
function KarmaRollFrameAwardButton_OnClick()

    local player = nks.RollList[KarmaRollFrame.selectedRoller][1];
    local value = -tonumber(KarmaRollFrameFinalKarma:GetText());
    if (KarmaRollFrameItem:GetText() == "") then
	-- If adding an item, reason can't be empty, it should contain item link
	Karma_message(nks.KMSG.ADDITEM);
	return;
    end
    Karma_message(player .. nks.KMSG.WINS .. KarmaRollFrameItem:GetText() .. nks.KMSG.USING .. nks.RollList[KarmaRollFrame.selectedRoller][3] .. nks.KMSG.TOTAL .. nks.RollList[KarmaRollFrame.selectedRoller][5] .. " (" .. KarmaRollFrameFinalKarma:GetText() .. nks.KMSG.PAYING .. ")" , nks.KARMA_SHOWTO_RAID);
    Karma_Add_Player(string.lower(player), value, KarmaRollFrameItem:GetText(), "I");
    KarmaRollFrameAwardButton:Disable();

end


function Karma_SetTip(link)
    KarmaTooltip:ClearLines()
    if (link ~= "" and string.match(link, "item:")) then
	KarmaTooltip:SetHyperlink(link)
    end
end


function Karma_SetItemRef(link, text, button)
    Karma_debug(3, "chatlink: link="..link);
    if ( not nks.Active ) then
        return
    end

    if (IsAltKeyDown()) then
	showWoWLoot(link);
    	Karma_SetTip(link);
    	Karma_update_rollframe(KarmaTooltip, text);
    end
end


function KarmaLootItem_OnClick(link)
    Karma_debug(3, "itemclick: link="..link);
    if ( not nks.Active ) then
        return
    end

    if (IsAltKeyDown()) then
	showWoWLoot(link)
    	Karma_SetTip(link)
    	Karma_update_rollframe(KarmaTooltip, link)
    end
end

function showWoWLoot(lootLink)
   -- Get the specs from wow-loot
   SendChatMessage("Bidding is now open for " .. lootLink, "RAID_WARNING")
   if WowLoot then
     local primary, secondary, tertiary = WowLoot:getAllFor(lootLink);
     if primary ~= "" then 
	     Karma_message("Primary - " .. primary, nks.KARMA_SHOWTO_RAID);
	     if secondary ~= "" then 
		     Karma_message("Secondary - " .. secondary, nks.KARMA_SHOWTO_RAID);
	     end
	     if tertiary ~= "" then
		     Karma_message("Tertiary - " .. tertiary, nks.KARMA_SHOWTO_RAID);
	     end
     else
	     Karma_message("The item " .. lootLink .. " is not in the WowLoot database yet");
     end
   end
end

function Karma_SetClassItem(tooltip)
    if (nks.OpenRoll) then
	KarmaRollFrameBaseCost:SetText(0);
    else
	local min_deduction = KarmaConfig["MIN_KARMA_NONCLASS_DEDUCTION"];
	local max_deduction = KarmaConfig["MAX_KARMA_NONCLASS_DEDUCTION"];
	local numlines = tooltip:NumLines();
	for n=1, numlines do
	    local text_field = getglobal(tooltip:GetName().."TextLeft".. n);
	    if (text_field) then
		local item_text = text_field:GetText();
		if (strfind(item_text, nks.KMSG.CLASSES)) then
		    min_deduction = KarmaConfig["MIN_KARMA_CLASS_DEDUCTION"];
		    max_deduction = KarmaConfig["MAX_KARMA_CLASS_DEDUCTION"];
		    break;
		end
	    end
	end
	KarmaRollFrameBaseCost:SetText(min_deduction);
    end
end


function Karma_Clearitem()
    KarmaRollFrameItem:SetText("");
    KarmaRollFrameBaseCost:SetText("0");
    KarmaRollFrameFinalKarma:SetText("0");
    KarmaRollList_Update();
end


function Karma_update_rollframe(tooltip, link)
    Karma_debug(3, "update rollframe: link="..link);	  

    -- there's one goofy bug here, if you have a tooltip up and click a link in chat, this
    -- will not search the tooltip for "Classes:" correctly (as it will hide the tooltip).
    if (string.match(link, "item:") and KarmaRollFrameItem:GetText() == "") then
	Karma_Clearitem();
	Karma_SetClassItem(tooltip);
	KarmaRollFrameItem:SetText(link);
    end
end


function KarmaRoll_OpenToggle()
    nks.OpenRoll = not nks.OpenRoll;
    -- KarmaRollFrameOpenCheckButton.checked = nks.OpenRoll;
    KarmaRollFrameOpenCheckButton:SetChecked(nks.OpenRoll);
end




NKS_Parse.lua
-- Ni Karma System (NKS) for raid loot distribution
-- The Ni Karma System was designed by Vuelhering (stef+nks @swcp.com) and Qed of Icecrown
-- 
-- Plugin coded by Mavios of Icecrown (althar @gmail.com).  Thanks Mavios, you rock.
-- Code maintenance and additional programming by Mavios and Vuelhering
-- Instructions for use at http://www.knights-who-say-ni.com/NKS
-- 
-- Copyright 2006-2009, Mavios and Vuelhering, Knights who say Ni, Icecrown
-- 
-- Permission granted for use, modification, and distribution provided:
-- 1. Any distributions include the original distribution in its entirety, OR a working URL to freely get the entire original distribution is clearly listed in the documentation.
-- 2. Any modified distributions clearly mark YOUR changes, or document the changes somehow.
-- 3. Any modified distributions MUST NOT imply in any way that it is an official upgrade version of this software (such as NKS+ or Enhanced Karma System, or probably anything with "NKS" or "Karma" in the name).  If you want your changes in the official distribution, write (stef+nks @swcp.com) and it might get included.
-- 4. No fee is charged for any distribution of this software (modified or original).
-- 
-- Snippets of code "borrowed" (fewer than 100 total lines) can merely include the URL http://www.knights-who-say-ni.com/NKS and credit for the code used.

-- Strip off a token
-- Return the lowercase token and remainder of line


Karma_GetToken = Ni_GetToken
Karma_StripTok = Ni_StripTok
Karma_debug = Ni_Debug

------------------------------------------------------------------------------------------
-- Command helper routines
-- Created by Tigerheart (http://www.wowwiki.com/HOWTO:_Extract_Info_from_a_Slash_Command)
 
function Karma_GetArgument(msg)
  if (msg) then
    local a,b=string.find(msg, "[^=]+");
    if (not ((a==nil) and (b==nil))) then
      local cmd=string.lower(string.sub(msg,a,b)); 
      return cmd, string.sub(msg, string.find(cmd,"$")+1);
    else  
      return "", "";
    end
  end
  return "", "";
end


localization.lua

-- Ni Karma System localization file
-- Got a translation?  Send it to me at stef+nks@ swcp.com
-- otherwise, I'm going to translate it using google, and that can't be good! :-)


-----------------------------------------------
-- Don't change this stuff here - same for all locales
-----------------------------------------------

BINDING_HEADER_KARMA = "Ni Karma System";

-- Only for local output, doesn't work in tells
local BLU = "|cff6666ff";
local GRY = "|cff999999";
local GRE = "|cff66cc33";
local RED = "|cffcc6666";
local ORN = "|cffcc9933";
local YEL = "|cffffff00";

-----------------------------------------------
-- English localization (Default)
-----------------------------------------------
BINDING_NAME_ROLLWINDOW = "Open/Close Roll Window";

nks.KMSG.LOADED = " loaded."
nks.KMSG.VERMISMATCH1 = RED .. "Ni Karma System Warning:" .. YEL .."  Need to upgrade from database version "
nks.KMSG.VERMISMATCH = RED .. "Ni Karma System Warning:\n" .. YEL .."Database version is outdated.  New databases may not be backward compatible with older program versions!\nClick OKAY to upgrade the database."
nks.KMSG.UPDATING = "Updating data from version "
--
-- silliness:
StaticPopupDialogs["NKSVERSION"] = {
	text = nks.KMSG.VERMISMATCH,
	button1 = OKAY,
	button2 = CANCEL,
	OnAccept = function()
		Karma_update_data();
	end,
	timeout = 60,
	whileDead = 1,
	hideOnEscape = 1
};


--commands
nks.KMSG.CUSE = "use"
nks.KMSG.CCREATE = "create"
nks.KMSG.COPTION = "option"
nks.KMSG.CCOMPACT = "compact"
nks.KMSG.CINFO = "info"
nks.KMSG.COFF = "off"
nks.KMSG.CROLL = "roll"
nks.KMSG.CSHOW = "show"
nks.KMSG.CADD = "add"
nks.KMSG.CSUB = "sub"
nks.KMSG.CADDITEM = "additem"
nks.KMSG.CSPAM = "spam"


-- standard bonus/nobonus
nks.KMSG.DECLARE = {}
nks.KMSG.REPLY = {}

nks.KMSG.DECLARE.BONUS = {"bonus", "bonsu"}
nks.KMSG.REPLY.BONUS = {"Your Karma of ", "will be added to your roll"}

nks.KMSG.DECLARE.NOBONUS = {"nobonus", "no bonus", "nobonsu", "no bonsu"}
nks.KMSG.REPLY.NOBONUS = {"Not using Karma on next roll"}

-- don't want to roll, but don't let it get sharded
nks.KMSG.DECLARE.NOSHARD = {"noshard", "shard", "no shard"}
nks.KMSG.REPLY.NOSHARD = {"Okay, this item will not be sharded"}

-- Additional modifiers ADDED TO END of bonus/nobonus:
--
-- don't want to roll against main specs, but will roll with bonus/nobonus
nks.KMSG.DECLARE.OFFSPEC = {"offspec", "off spec"}
nks.KMSG.REPLY.OFFSPEC = {"Declared as offspec (passing to any main specs)"}

-- might want to pass to someone if they're rolling, so don't drop it on me immediately
nks.KMSG.DECLARE.PASS = {"pass"}
nks.KMSG.REPLY.PASS = {"Will check if you want to pass, after the roll"}

nks.KMSG.BONUS = "bonus"
nks.KMSG.NOBONUS = "nobonus"
-- for splitting the words... allow "no bonus" in addition to "nobonus"
nks.KMSG.NOBONUS1 = "no"
nks.KMSG.NOBONUS2 = "bonus"


nks.KMSG.REPLYBONUS1 = "Your Karma of "
nks.KMSG.REPLYBONUS2 = " will be added to your roll"
nks.KMSG.REPLYNOBONUS = "Not using Karma on next roll"

-- command arguments, including player commands
nks.KMSG.HELP = "help"
nks.KMSG.ALL = "all"
nks.KMSG.HISTORY = "history"
nks.KMSG.ITEMS = "items"
nks.KMSG.KARMA = "karma"

nks.KMSG.HELP2 = "Fields in [brackets] are optional\nThe database name is case sensitive";
nks.KMSG.HELP3 = "/KM command\nCommands:\n  HELP\n  OFF\n  INFO\n  ROLL\n  CREATE database\n  USE database\n  SHOW (playername\|class\|ALL) [KARMA\|ITEMS\|HISTORY] [TO\|RD]\n  ADD [-]# (ALL, player) [reason]\n  ADDITEM [-]# (ALL, player) itemlink [comment]\n  OPTION SHOW \| setting=ON\|OFF\|#\n  COMPACT #ofdays\n  SPAM";

-- player messages when sending "km <command>" tells to the loot master
nks.KMSG.PLAYER_HELP1 = "Ni Karma System Help"
nks.KMSG.PLAYER_HELP2 = "Fields in [brackets] are optional"
nks.KMSG.PLAYER_HELP3 = "/t <loot_master> <command>\n  km show\n  km show [karma [class/player]]\n  km show history\n  km show items [class/player\]\n  km help\n  Ex: To get your karma, use \"km show\".  To get a list of the warrior karma in the raid, use \"km show karma warrior\""

-- the following is a FAST description of the system, and can be spammed for new players with /km spam.
nks.KMSG.SPAM = "The Ni Karma System adds your Karma score to a /roll (1-100 + bonus) when you send me a tell of \"" .. nks.KMSG.BONUS .. "\" and does a normal roll (1-100) when you send me a tell of \"" .. nks.KMSG.NOBONUS .. "\".\nWhen asked to declare on items, send me only \"" .. nks.KMSG.BONUS .. "\" or \"" .. nks.KMSG.NOBONUS .. "\", and /roll when told.\nOnly those within 50 karma of the highest \"bonus\" score can roll.\nIf you win and are using Karma bonus, you lose half.  There is no loss if you don't use bonus, or don't win, but class-specific items have a min/max loss, usually 25/100 pts."

nks.KMSG.NORAID = "No active raid";
nks.KMSG.BADCOMMAND = "Invalid Command";

nks.KMSG.YOU = "You"
nks.KMSG.PLAYER = "Player: "
nks.KMSG.NOHISTORY = " is not in raid history"
nks.KMSG.CURRENT1 = "Your current Karma: "
nks.KMSG.CURRENT2 = "Current Karma of "

nks.KMSG.ADDNOPOINTS1 = "You must specify amount of karma to ADD"
nks.KMSG.ADDNOPOINTS2 = "Karma to ADD must be a number (positive or negative)"
nks.KMSG.OFFLINELIST = "Offline (no karma added): "

nks.KMSG.ADDITEM = "You must specify an item"
nks.KMSG.USERAID = "You must specify a database"
nks.KMSG.NOTFOUND = "Database not found.  (Check your spelling and capitalization, or " .. BLU .. "/km create <dbname>"..YEL.." to create a new database)"
nks.KMSG.EXISTS = "Database already exists and will not be created again"
nks.KMSG.CREATED = "Creating new database ("
nks.KMSG.USINGRAID = "Karma running on database ("
nks.KMSG.DISABLED = "Karma is off.  " .. BLU .. "/km use <Database>" .. YEL .. " to load one, or " .. BLU .. "/km create <Database>" .. YEL .. " to create a new database"

nks.KMSG.ON = "ON" -- uppercase
nks.KMSG.OFF = "OFF"

-- options are ONLY the words, do not use for table indices
nks.KMSG.SHOW_WHISPERS = "SHOW_WHISPERS"
nks.KMSG.NOTIFY_ON_CHANGE = "NOTIFY_ON_CHANGE"
nks.KMSG.ALLOW_NEGATIVE_KARMA = "ALLOW_NEGATIVE_KARMA"
nks.KMSG.MIN_KARMA_CLASS_DEDUCTION = "MIN_KARMA_CLASS_DEDUCTION"
nks.KMSG.MAX_KARMA_CLASS_DEDUCTION = "MAX_KARMA_CLASS_DEDUCTION"
nks.KMSG.MIN_KARMA_NONCLASS_DEDUCTION = "MIN_KARMA_NONCLASS_DEDUCTION"
nks.KMSG.MAX_KARMA_NONCLASS_DEDUCTION = "MAX_KARMA_NONCLASS_DEDUCTION"
nks.KMSG.KARMA_ROUNDING = "KARMA_ROUNDING"
nks.KMSG.BADOPTION = "Unknown option"

nks.KMSG.NORECORD = "No record of you in current database"

nks.KMSG.RECENT1 = "Your recent events:"
nks.KMSG.RECENT2 = "Recent events for "

nks.KMSG.DEDUCTED = " karma deducted for "
nks.KMSG.ADDED = " karma added for "
nks.KMSG.COST = " karma for "
nks.KMSG.BADSHOW = "Invalid SHOW command"

nks.KMSG.COMPACTING = "Compacting old entries for "
nks.KMSG.COMPACTED = "Compact complete"
nks.KMSG.ROLLED = " roll "
nks.KMSG.REROLLED = RED .. " rerolled." .. YEL
nks.KMSG.NOTINRAID = " tried to join roll list but is not in the raid"
nks.KMSG.WINS = " is the winner of "
nks.KMSG.USING = " using "
nks.KMSG.TOTAL = " karma for a total of "
nks.KMSG.PAYING = " karma spent"

-- compacted entry
nks.KMSG.OLDENTRIES = "Karma From Old Entries"

-- OLD: not really a message, but will have localization issues
-- OLD: I'm using '|' for string searching, so it should have those at each end of each word
-- OLD: nks.KMSG.allclasses = "|druid|hunter|mage|paladin|priest|rogue|shaman|warlock|warrior|";


--
-- nks.KMSG.CLASS.localname = "db_class"
-- You can have multiple localnames and aliases if you like.
--
nks.KMSG.CLASS = { }

-- This section has two parts.  First you must have the all classes as returned by GetRaidRosterInfo
-- mapped to the english version.
-- Next, you can have aliases, mapping to the appropriate english version.
-- ex:
-- ["hexenmeister"] = "Warlock",
-- ["hexenmeisterin"] = "Warlock",
-- ["warlock"] = "Warlock"
-- ["dk"] = "Death Knight"
--
-- I didn't add all the languages here, because someone on the US client might name their warlock "Hexenmeister"
-- Add the local names as needed based on GetLocale()
--
nks.KMSG.CLASS.druid = "Druid"
nks.KMSG.CLASS.hunter = "Hunter"
nks.KMSG.CLASS.mage = "Mage"
nks.KMSG.CLASS.paladin = "Paladin"
nks.KMSG.CLASS.priest = "Priest"
nks.KMSG.CLASS.rogue = "Rogue"
nks.KMSG.CLASS.shaman = "Shaman"
nks.KMSG.CLASS.warlock = "Warlock"
nks.KMSG.CLASS.warrior = "Warrior"
nks.KMSG.CLASS["death knight"] = "Death Knight"
nks.KMSG.CLASS.deathknight = nks.KMSG.CLASS["death knight"]

-- aliases for the command /km show <class>
nks.KMSG.CLASS.dk = nks.KMSG.CLASS["death knight"]

-- roll window
nks.KMSG.ROLL = { };
nks.KMSG.ROLL.MIN = "Min Karma:"
nks.KMSG.ROLL.MAX = "Max"

-- system output
nks.KMSG.SYS = { };
nks.KMSG.SYS.ROLLS = "rolls"

nks.KMSG.CLASSES = "Classes:"


-----------------------------------------------
-- German localization
-----------------------------------------------
if ( GetLocale() == "deDE" ) then

-- system output
nks.KMSG.SYS.ROLLS = "w\195\188rfelt. Ergebnis:"

-- German names from Alexander 'Bl4ckSh33p' Spielvogel and his brother ESN
-- Thanks!
-- English names from above should also work.
nks.KMSG.CLASS = {
	["hexenmeister"] = nks.KMSG.CLASS.warlock,
	["hexenmeisterin"] = nks.KMSG.CLASS.warlock,
	["krieger"] = nks.KMSG.CLASS.warrior,
	["kriegerin"] = nks.KMSG.CLASS.warrior,
	["j\195\164ger"] = nks.KMSG.CLASS.hunter,
	["j\195\164gerin"] = nks.KMSG.CLASS.hunter,
	["magier"] = nks.KMSG.CLASS.mage,
	["magierin"] = nks.KMSG.CLASS.mage,
	["priester"] = nks.KMSG.CLASS.priest,
	["priesterin"] = nks.KMSG.CLASS.priest,
	["druide"] = nks.KMSG.CLASS.druid,
	["druidin"] = nks.KMSG.CLASS.druid,
	["paladin"] = nks.KMSG.CLASS.paladin,
	["schamane"] = nks.KMSG.CLASS.shaman,
	["schamanin"] = nks.KMSG.CLASS.shaman,
	["schurke"] = nks.KMSG.CLASS.rogue,
	["schurkin"] = nks.KMSG.CLASS.rogue,
	["todesritter"] = nks.KMSG.CLASS.deathknight,
}


-----------------------------------------------
-- French localization
-----------------------------------------------
elseif ( GetLocale() == "frFR" ) then


end

